### PR TITLE
Icing/client certs

### PR DESCRIPTION
**Pull request by icing**
* add `rustls_root_cert_store_builder` and `rustls_root_cert_store` to build and add to a server config
* add `rustls_server_session_get_peer_certificate` to retrieve the peer certificate (chain) from a session
* add `rustls_certificate` in the API without any accessors yet due to missing support

### CODE REVIEW

**Review by jsha — CHANGES_REQUESTED** [review]
This is generally looking quite good! I think one thing to still figure out is whether to represent `rustls_root_cert_store` as a `Box<RootCertStore>` or an `Arc`. Currently I think nothing in rustls takes an `Arc<RootCertStore>`. The places in this PR that use the `Arc`-ness of it wind up immediately cloning the inner Vec, which means we're not really taking advantage of the `Arc`-ness.

I think to get the behavior you want, we should take advantage of the `Arc`-ness in:

    new(client_cert_verifier: Arc<dyn ClientCertVerifier>) -> ServerConfig

That is, we should export bindings for `AllowAnyAuthenticatedClient` and `AllowAnyAnonymousOrAuthenticatedClient`, and those bindings should actually return `Arc<dyn ClientCertVerifier>`. Then, to create a new server config cheaply, we can clone that arc (which just increments the refcount), and pass the resulting pointer to ServerConfig::new().

If we do that, then it makes sense to make `rustls_root_cert_store` a Box instead of an Arc, because it winds up being fully owned by the `AllowAnyAuthenticatedClient` or `AllowAnyAnonymousOrAuthenticatedClient` we give it to.

File: src/cipher.rs [hunk 1]
```diff
@@ -110,6 +126,30 @@ pub extern "C" fn rustls_certified_key_build(
...
+    key: *const rustls_certified_key,
+    i: size_t,
+) -> *const rustls_certificate {
+    ffi_panic_boundary! {
+        let certified_key: Arc<CertifiedKey> = unsafe {
+            match (key as *const CertifiedKey).as_ref() {
+                Some(c) => arc_with_incref_from_raw(c),
+                None => return null(),
+            }
+        };
```
[the remark this answers is not in the document]
Comment (icing) [reply src/cipher.rs:145]: Replacing this with `try_ref_from_ptr` does not work. The macro returns `&CertifiedKey`.

**Comment by icing** [comment]
> Currently I think nothing in rustls takes an `Arc<RootCertStore>`. The places in this PR that use the `Arc`-ness of it wind up immediately cloning the inner Vec, which means we're not really taking advantage of the `Arc`-ness.

True. I originally had the hope to convince rustls to provide a Verifier with an `Arc` store.

> If we do that, then it makes sense to make `rustls_root_cert_store` a Box instead of an Arc, because it winds up being fully owned by the `AllowAnyAuthenticatedClient` or `AllowAnyAnonymousOrAuthenticatedClient` we give it to.

Yes, we can do that. But that means the application has to do the "paperwork" of keeping track of tuples `(storeFile, mandatory)` in order to have reuse. This design does not really make sense from the app's point of view. Having a `ClientVerifier` inside rustls is quite good, but it is nothing the application cares about. 

The application needs to assemble `ServerConfig`s from root stores and mandatory flags, which are configured and can come in any variation. While it is unlikely that someone configures the same CA store with different flags, the application code has still to be able to handle that.

At this stage, I would suggest making the `Arc` into a `Box`, but leave the rest of the PR. The only drawback is then that we have duplicate certificate DER data in memory *if* the same CA file is configured for several servers. Since these should not be that large, I guess we can live with that.

**Comment by jsha** [comment]
> the application has to do the "paperwork" of keeping track of tuples (storeFile, mandatory) in order to have reuse.

I don't fully understand this. I think with the current PR, the application has to do paperwork of keeping track of storeFile -> *rustls_root_cert_store. With my proposal the application would have to do paperwork of keeping track of (storeFile, mandatory) -> *rustls_client_cert_verifier. Does using a tuple vs a string make the paperwork that much worse?

**Comment by icing** [comment]
> > the application has to do the "paperwork" of keeping track of tuples (storeFile, mandatory) in order to have reuse.
> 
> I don't fully understand this. I think with the current PR, the application has to do paperwork of keeping track of storeFile -> *rustls_root_cert_store. With my proposal the application would have to do paperwork of keeping track of (storeFile, mandatory) -> *rustls_client_cert_verifier. Does using a tuple vs a string make the paperwork that much worse?

As I said before, this can be done. But it will load `fileA` twice if the server is configured with `(fileA, true)` and `(fileA, false)`. The nails on the chalkboard are just in my head...I will survive this. ;-)

If you want the 2 rustls verifier implementations exposed in the crustls API, let me know and I can do that tomorrow.

**Comment by jsha** [comment]
> If you want the 2 rustls verifier implementations exposed in the crustls API, let me know and I can do that tomorrow.

Yes, please. :-)

**Comment by icing** [comment]
Done. I tried to keep it at one `rustls_client_cert_verifier` that keeps an `Arc<dyn ClientCertVerifier>`. But on convincing rustc to compile it, it only panic'ed at runtime.

**Review by jsha — APPROVED** [review]
Great! I made some edits, which you can see in https://github.com/abetterinternet/crustls/pull/98/commits/7f5da4663f86f5395686ce0849957927c558ab28.

I'd be curious to see the code and/or panic message you got when trying the "just one rustls_client_cert_verifier" variation, if you still have them around.

**Pull request merged**